How to Integrate Varnish Cache with Nginx for both SSL and plain HTTP WordPress Websites, on Debian 8 (Jessie)

Fred-Zweig opened the pull request:
This tutorial presents the most effective method of integrating Varnish with Nginx to serve cached WordPress content, for both SSL and plain HTTP websites.

alexfornuto commented:
Please squash your commits: https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

Fred-Zweig commented:
I accidentally closed this branch.

Fred-Zweig commented:
I reopened the branch. My apologies.

Fred-Zweig commented:
Hi everybody!  

To be honest I'm a bit disappointed that nobody seems to have read my tutorial during these 3 months since I uploaded it. I put much effort in writing the most clear and detailed instructions, giving this guide the best form possible. It's not a sketch. It's carefully written, formatted and ready for review. 

Who doesn't need a web server ? So, why should you use Apache since from my experience Apache uses about one hundred times more RAM than Nginx. So, if Nginx does what Apache does (and even more) by using so much less resources, is it an exaggeration to call Nginx a better web server ? It's not only my personal experience, many published statistics support the superiority of Nginx. And who doesn't need a CMS ? And why should you use a paid proprietary CMS if you can do the same things with a free open source CMS like WordPress ? So, if you use WordPress you know you need a caching solution which drastically improves the speed of any website. I personally tried caching plugins such as 'W3 Total Cache' and 'WP Super Cache' and they cannot be even compared to Varnish. In spite of what some say, Varnish makes all caching plugins superfluous. So, you use WordPress and Varnish but what can you do about the SSL websites, since Varnish caches only plain HTTP content. So, you need a server to decrypt the SSL traffic. But why should you use a third server for decryption, like Pound for example, since you can use Nginx both for decrypting SSL traffic and for serving HTTP content ? By using only Nginx and Varnish you reduce setup complexity and the number of possible failure points. 

So, I think the only reason why still few people use this setup is that they don't know it is possible and also don't know how to configure Nginx and Varnish to make them the most efficient and speedy setup possible for both HTTP and HTTPS traffic.

alexfornuto commented:
Hi @Fred-Zweig. I really do apologize for the delay in reviewing this submission. We're a small team, and have been in a bit over our head with the amount of guides that have been submitting lately. We're doing our best to process them in the order in which they were received, and yours is reaching the top of the list.

I have to ask, how much of what's in this guide is information already in our existing [varnish guide](https://www.linode.com/docs/websites/varnish/getting-started-with-varnish-cache)? Would it make sense to set this guide up as a follow-up explaining how to take the Varnish install described in the existing guide and configure it for the Nginx configuration described in yours?

Fred-Zweig commented:
Thank you for your reply. The [guide](https://www.linode.com/docs/websites/varnish/getting-started-with-varnish-cache) that you mentioned describes just the general configuration of Varnish 3.0 for a plain HTTP website served by Apache, on Ubuntu 12.04. My guide presents the configuration of Varnish 4.0 (different from the 3.0 version) so that it can be used to cache both plain HTTP and SSL websites served by Nginx, on Debian 8 (that uses the default ‘systemd’ init system). It explains how to configure Varnish to make it cache Wordpress websites and it describes in detail how to install and configure Nginx so as to make it simultaneously decrypt SSL traffic and serve HTTP content, in conjunction with Varnish. So, my guide’s topic and coverage are very different from those of the published guide. I wouldn’t think of it as a follow-up of the published guide, but rather as an up-to-date guide about installing and configuring Varnish (the best existing caching solution) to make it work with the most efficient web server which is Nginx, for all types of Wordpress websites, be they plain HTTP or SSL websites.

The main focus of my guide, which I don’t find the words to stress enough here, is the idea that with the right setup, Varnish can be used not only for HTTP traffic, but for SSL traffic as well. We all know that by default Varnish cannot cache SSL content. The SSL/TLS protocol is becoming increasingly important and it is even a ranking factor for Google, so what’s the use of implementing Varnish if it can’t be used for SSL websites ? By using a SSL proxy like Pound to decrypt the SSL traffic, Varnish can be used for SSL websites also. But why use a third server like Pound, since we can harness the power and versatility of Nginx to make it both a SSL proxy and a backend web server ?
 
Since last year there is a proprietary version of Varnish called ‘Varnish Plus’ which offers SSL support. ‘Varnish Plus’ is offered together with technical support in a subscription package that starts from [$31000](https://www.varnish-software.com/pricing/varnish-plus) for a three-node cluster.

Speaking about SSL proxies that are generally used with Varnish to make it handle SSL traffic, Per Buer, CTO and founder of ‘Varnish Software’, the company that sells ‘Varnish Plus’, aknowledges in one of his articles that ‘[There are quite a few SSL proxies out there and most of them play quite nicely with Varnish.](https://info.varnish-software.com/blog/flying-pigs-and-ssl-varnish-cache-plus)’  So, since a Varnish developer acknowledges that Varnish can be used effectively with a SSL proxy, why not use Nginx as a SSL proxy, as I describe in my guide ? Of course, ‘Varnish Plus’ may be useful for very complex setups with multiple physical servers in multiple locations, and it may be in some situations faster by a few milliseconds than my setup, but for the majority of cache implementation needs I think my solution is more than enough. It's simple, elegant and it saves a lot of money.

brewn commented:
Hi @Fred-Zweig  -

Thanks again for your submission - I've been making some edits and testing out this guide. One of the things I've noticed is that the configuration (under /etc/nginx/sites-available/default) appears to configure both http and https on port 8080. Unless I'm mistaken, nginx will view this as a syntax error without some pretty serious modifications. 

Additionally, the use of "example-over-ssl.com" vs. "example-over-http.com" is likely going to be confusing, as in actuality the user is going to be entering the same domain name for each server block, assuming (as this guide does) that we're setting this up for a single domain. I wanted to run this my you to see if I've misunderstood the purpose of this naming scheme - if it's simply a stylistic choice, I can make the necessary changes myself. 

I've included some gists with the error messages, as well as the config files themselves. I've replaced the test domain I'm using with "example.com," and I've removed the actual IP address:

Error message (varnish):
https://gist.github.com/brewn/f86251e4967921a9998dc88f5ec9905e

/etc/varnish/custom.vcl:
https://gist.github.com/brewn/139afb0b50d4c2885763de95e7aa7a33

/etc/default/varnish:
https://gist.github.com/brewn/cb82f4fbff3055ffb07fc2c4d5677a12

I've also included the nginx server block configuration, and the output of 'nginx -t.' The syntax check seems to find an issue with most lines after the 'Index' directive in the server block for SSL on 8080. 

/etc/nginx/sites-available/default:

https://gist.github.com/brewn/1ef4a8e17e207dcf533130e699c2db8a

Output from nginx -t:

nginx: [emerg] invalid number of arguments in "root" directive in /etc/nginx/sites-enabled/default:86
nginx: configuration file /etc/nginx/nginx.conf test failed

Line 86:

 86	   index index.php index.html index.htm;

I feel like it's possible I'm missing a small syntax issue here, but I've tried to copy entirely from the existing guide as much as possible. A few of the directories used are different , but I don't believe these would be resulting in a syntax error.

Fred-Zweig commented:
@brewn

You seriously misunderstood the whole setup. It’s really upsetting to see such objections that prove you haven’t read the guide carefully. I’ll dismantle your objections one by one and I’ll explain the setup once more but I won’t repeat these explanations afterwards, so if you want to make other objections you’ll have to read the guide and these explanations carefully:

1. First you mentioned a possible problem with the configuration under "/etc/nginx/sites-available/default". You wanted to say "/etc/nginx/sites-enabled/default". Of course that is just a simlink to the 'default' in 'sites-available', but let’s be accurate with the terms used.

Then you say that the configuration in the 'default' file "appears to configure both http and https on port 8080" and that "the use of "example-over-ssl.com" vs. "example-over-http.com" is likely going to be confusing,". These statements prove you didn't understand the setup at all. I'll explain it again. It goes like this:
 
In a real life situation on your server you have multiple websites served by the same instance of Nginx. Some of these websites are served over http and some of them are served over https. But these websites are different from one another. You either serve a website over http, or you serve it over https. You don't serve the same website simultaneously over http and https. This is absurd. The idea in itself is absurd, not to mention that Nginx won’t start. Yet this is precisely what you are doing in the configuration posted by you here:
 
https://gist.github.com/brewn/1ef4a8e17e207dcf533130e699c2db8a

You try to serve the same website "www.example.com" over http on port 8080 and also over https on port 443. No offense but this is nonsense. The guide speaks about two completely different websites: "www.example-over-http.com" and "www.example-over-ssl.com". These websites are different, and this is precisely why I named them this way. If they had been the same website they would have been both called "www.example.com" as you mistakenly called yours. It’s obvious enough that they are different websites. I’m wondering how you could think to configure the same website in two server blocks so as to serve it simultaneously over http and https.

The fact that "www.example-over-ssl.com" appears in two server blocks, the first with port 443 and the second with port 8080 shouldn’t be confusing if you read carefully the "Our setup" section of the guide: For https traffic frontend Nginx acts as a proxy, decrypting traffic received on port 443 and sending it unencrypted on port 80 to Varnish, as this line of the server block clearly shows: "proxy_pass http://127.0.0.1:80;"  But what happens afterwards ? Afterwards Varnish receives the request from frontend Nginx and if it finds the content in cache it delivers it back to frontend Nginx; if it doesn’t find it in cache it will ask backend Nginx on port 8080 for that specific content. So the "www.example-over-ssl.com" website has to have a second server block with port 8080, as if it were a plain http website, so as to enable backend Nginx to serve the required content to Varnish. 

So, the names of the two different websites are carefully chosen and shouldn’t be changed.

2. You also say about Nginx that "The syntax check seems to find an issue with most lines after the 'Index' directive in the server block for SSL on 8080". No wonder that the syntax check finds an issue with the lines after 'Index', since before 'Index' you forgot to put a semicolon. Look at your own configuration on line 66:

https://gist.github.com/brewn/1ef4a8e17e207dcf533130e699c2db8a

My guide doesn’t have that mistake. And even if you correct that mistake in your configuration, Nginx still won’t start because, as I said, you can’t serve the same website "www.example.com" simultaneously over http and https.

3. You also say you had troubles starting Varnish. No wonder you had problems starting Varnish since you made the following mistakes in your configuration files, which mistakes are not present in my quide:

In your "/etc/default/varnish" you forgot to place a backslash ( \ ) after "-T localhost:6082" on line 49 of your configuration file:

https://gist.github.com/brewn/cb82f4fbff3055ffb07fc2c4d5677a12 

Also in your "/etc/varnish/custom.vcl" you forgot to place a closing curly bracket ( } ) at the end of the file. Look on line 79 of your configuration file:

https://gist.github.com/brewn/139afb0b50d4c2885763de95e7aa7a33

Also in your  "/etc/varnish/custom.vcl" on line 16, the https website has to be preceded by "www" if you really want to follow the guide, though this is not the source of a Varnish error.

I also cannot know if you made the correct settings in "/lib/systemd/system/varnish.service", following the steps described in the guide (steps 22 to 25).

In "/etc/varnish/custom.vcl" there is nothing wrong with this section:

 if (req.url ~ "wp-admin|wp-login") {
    return (pass);
 }

I tested Varnish again and again with the "/etc/varnish/custom.vcl" configuration described in the guide and the result is always the same: it works. It starts with no problems.

So, all your objections are groundless.

@alexfornuto and @pbzona
 
Why don’t you test my guide yourselves when you have the time. I’ve been waiting since April and I'll wait more if necessary. It's a real pain and loss of time to defend my guide against poorly thought objections. I won’t do it in the future. If someone doesn’t want to make the effort to read the guide carefully and understand it, he shouldn’t attempt to make corrections, and I shouldn’t be put in the situation to explain again and again what is already in the guide.

alexfornuto commented:
@Fred-Zweig 

Thanks for your detailed response. @brewn is a member of our team, and just as qualified to handle this submission as any other member. We all do our best to test these guides carefully. If we make mistakes, it is because we are human. It may also indicate the complex nature of the document in question, and show how easily a reader of the finished product might make the same mistake. 

Because we strive for excellence in any guide we ultimately publish, we perform several thorough tests of each submission, and ask for author cooperation in resolving any concerns that arise. If you feel that this process is too much of a "pain and loss of time", then we can close this pull request.

As we've documented in the contributing page, the bounty program itself is on hold so that we can process the guides currently waiting, and then launch a new platform that will make it easier for authors to see the end product on their own system, and have topics discussed and approved before they need to be written out.

I encourage you to be patient and understanding as we review and test your submission. If you can do that, @brewn will review your last update and continue testing the guide.

Fred-Zweig commented:
@alexfornuto 
Thank you for your reply. I'm sorry if I've sounded harsh. I didn't want to offend anybody. I'll wait. I'm ready to acknowledge my mistakes if there are any, but so far there are none.

Fred-Zweig commented:
I edited my previous 'defense' post and I removed the harsh part leaving only the objective facts.

brewn commented:
@Fred-Zweig -

Your input is appreciated - as I said, I certainly didn't intend to rule out the possibility of a syntactical issue on my end, and I apologize if my questions at any time came across as accusatory. 

I think I'm a little unclear on one of your points here, and I'd just like to make sure we're on the same page before I continue working on this guide:

> Some of these websites are served over http and some of them are served over https. But these websites are different from one another. You either serve a website over http, or you serve it over https. You don't serve the same website simultaneously over http and https.

I can't help but feel like I'm misreading you here, as It's quite common to provide access to the same domain over both HTTP and HTTPS. Admittedly I am not as familiar with the process once Varnish is involved. So to be clear - for this guide, you're serving one domain over HTTP, and entirely different one over HTTPS?

Fred-Zweig commented:
@brewn 

Yes, the two example websites in my guide are different websites. They are two different domains. One is served over http and the other over https. To make things more obvious, I think I'll  rename the "www.example-over-http.com" website. I'll name it: "www.different-example-over-http.com", so as to make it more clear that it is completely different from "www.example-over-ssl.com".

And indeed, you are right, there are cases in which some pages of a website are served over http, to gain speed, and other pages are served over https, like checkout pages on e-commerce websites, for which the secure connection is more important. So, in such a case, the same website, which means the same domain, is served both over http and over https, but the pages that are served over http are always served over http and the pages served over https are always served over https. The same pages are not served simultaneously over http and https. This was the idea that I wanted to stress in my previous post. I pointed out this idea because the Nginx configuration that you posted, appeared to try to serve the same whole website, with all its pages, **simultaneously** over http and over https. 

When I wrote the guide I didn't have in mind the case in which some pages of a website would be served over http and other pages over https. Such cases are more complex and less frequent. I described the general case in which you serve the whole website either over http or over https. In fact this is the most logical and efficient setup when Varnish is involved.

When Varnish is added to the setup, things are much simpler if the whole website is served either over http or https. If you try to serve some pages of a website over http and other pages of the same website over https, the setup becomes much more complicated and this complexity would be pointless, since the only reason one would serve some pages of a https website over http, would be to gain speed, but this is precisely why Varnish is added to this setup, to provide speed. And it really succeeds to improve speed drastically for http and also for https websites. So, if Varnish and Nginx are properly configured, all websites, be they http or https websites, will be served so quickly that it would not make sense to split a website into two parts: one part served over http and another over https. Whenever connection security becomes important, the whole website can be served over https. And if security is not important, such as for a simple presentation website, for which the owner doesn't want to buy a SSL certificate and renew it periodically, then a plain http website would be enough. 

I'm aware that there may be some exotic cases in which the same page of a website is served both over http and over https depending on the IP or IP geolocation of the client requesting the page or depending on other criteria, but even in this situation that page won't be delivered to the same client simultaneously over http and over https. Anyway, these situations are so rare that I don't think a Linode guide on such a topic would be of real use to the majority of readers.

Fred-Zweig commented:
I changed "www.example-over-http.com" to "www.different-example-over-http.com", so as to make it even more obvious that it's different from "www.example-over-ssl.com".

I also added a warning about the "WooCommerce Recently Viewed" widget which must be disabled if we want to cache WooCommerce product pages.

Fred-Zweig commented:
When testing server behavior under heavy load with JMeter, I got very good results even on a server with medium specifications (2 cores CPU at 2.4 GHz, 3.3 GB of RAM, 80 GB HDD):

- for the https test website: when 1000 concurent anonymous users were connected, average response time was 96 miliseconds and the server throughput was about 10 Mbit/s, with 0 % errors.
- for the http test website: when 1000 concurent anonymous users were connected, average response time was 7 miliseconds and the server throughput was about 26 Mbit/s, with 0 % errors.

Of course, server response time and throughput depend on many variables such as server hardware specifications, website complexity, whether users are anonymous or logged in users, etc. Nevertheless, since I used WordPress websites with medium complexity and a server with medium specifications, I think my results are relevant and prove that the setup that I describe in my guide is very efficient for both protocols: http and https.

brewn commented:
@Fred-Zweig Thanks. I've been working on a version of this guide and I'm ready to push my changes, but it's going to result in a merge conflict since commits have been made after I originally fetched this PR. Would it be possible to roll back your previous commits? I've added a clarification to the guide that should sort out the example-over-http/https confusion, and we can revisit the Woocommerce warning before we publish.

Fred-Zweig commented:
@brewn In the "Our setup" section I added the most clear details myself:

" We will use as example two different websites: "www.example-over-ssl.com" will be the website delivered over the HTTPS protocol and "www.different-example-over-http.com" will be a different website delivered over the HTTP protocol. So, if you want to use Nginx to serve both HTTPS and HTTP websites and benefit from Varnish caching for all of them, you should configure your HTTPS websites like "www.example-over-ssl.com" and your HTTP websites like "www.different-example-over-http.com". "

brewn commented:
@Fred-Zweig would you be able to roll back the changes from after June 5th? If you can hold on to them on a local branch, we can consider applying them later - but we have numerous changes in our version that will cause a merge conflict, since this PR has been updated since we began work on it.

Fred-Zweig commented:
@brewn Rolling back to previous commits and changing history is risky and I really don't see the necessity. If you have any addition or modification to make please first state it here accompanied with the arguments for that change. I brought this guide to the most clear form possible. I doesn't need any additional clarification. If you believe otherwise, please state your arguments here to discuss them.

brewn commented:
@Fred-Zweig The majority of the changes are stylistic - mostly small revisions to wording. It wouldn't be possible to enumerate them individually, and it's much easier to view these sort of changes in a PR. 

I can try to resolve this merge conflict myself, but it should be understood that our acceptance of this guide is going to rely upon a review process where we will make changes to bring it in line with the style of our documentation.

Fred-Zweig commented:
@brewn I can definitely agree with stylistic changes, like different wording. The merge conflict may be due to the fact that I closed this pull request on June 6th when trying to squash my previous commits. If you cannot solve the merge conflict I can operate all the changes myself in the last version of the guide if you can send me the corrected text as attachment to an email. From my previous experience with github I don't think a rollback to a version prior to June 5th will work.

alexfornuto commented:
@Fred-Zweig The merge conflict was created because you were making changes while we were editing. Generally speaking, you want to avoid working on the same file at the same time. Also, we had to rename the file to match our convention, so your changes show as being part of a separate guide. Because of this, we can't even see a manual line-by-line conflict, which would be possible to merge cleanly. 

@brewn will send you a Pull Request with his edits shortly. The easiest way to merge it will probably be to roll back your two last commits. If you'd prefer to merge manually, that's your prerogative, but I wouldn't bother. This guide is going to go through several more review steps, so the copy is likely to change anyway. The current steps revolve simply around making sure the guide works and is understandable.

Fred-Zweig commented:
@alexfornuto Ok. I'll roll back my last three commits. This way the guide will be reverted to the form it had before September 19. And then I'll wait for the pull request with stylistic changes.

brewn commented:
@Fred-Zweig Here's that PR:

https://github.com/Fred-Zweig/docs/pull/1

Fred-Zweig commented:
@brewn Thank you. I rolled back my last three commits and merged your branch successfully. All the changes that you made in your branch are now incorporated in the guide. I had to do other minor formatting changes but now the guide looks good.

Fred-Zweig commented:
I also added the warning about the "WooCommerce Recently Viewed" widget.

brewn commented:
@Fred-Zweig Thanks. I'll pass this along for the next round of testing. We should have more updates for you soon. Just to avoid another merge conflict, please don't push any new commits in the meantime.

Fred-Zweig commented:
@brewn Ok. I won't make any changes to the guide until notified.

As it is now the guide has the title: "How to Use Varnish with Nginx to Serve Cached Wordpress Content over SSL on Debian 8". Wouldn't it be more appropriate to call it: "How to Use Varnish with Nginx to Serve Cached Wordpress Content over SSL and plain HTTP on Debian 8", since the guide also describes how to serve http websites ? The guide is equally dedicated to http and https websites. A user who might be interested to configure Varnish and Nginx just for a http website may read the title and think that the guide doesn't address http websites, so he will browse away thinking the guide is not for him.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+				 	-T localhost:6082 \
+				 	-f /etc/varnish/custom.vcl \
+				 	-S /etc/varnish/secret \
+				 	-s malloc,500M"
```
Two quick questions here. 

1) Now that the minimum Linode plan has 2 gigabytes of RAM, can/should we update this variable?
2) By default, the configuration file uses a lowercase `m`. Is there any difference as far as Varnish is concerned, or any other reason to change it here?

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+
+	This will set Varnish to listen on port 80, and will instruct it to use the `custom.vcl` configuration file. If you make the configuration changes in the `default.vcl` file, a future update to Varnish may overwrite it. Save and exit the file.
+
+	The `-s malloc,500M` line sets the maximum amount of RAM that will be used by Varnish to store content. This value can be adjusted to suit your needs, taking into account the server's total RAM along with the size and expected traffic of your website. For example, on a system with 4 GB of RAM , you can allocate 1 or 2 or 3 GB to Varnish but obviously not all the system's RAM. To allocate 1 GB yo ...
```
As with line 83, is there a difference between `g` and `G` as far as Varnish is concerned?

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+	
+		acl purger {
+		"localhost";
+		"your_server_ip";
```
Are these dummy variables that should be replaced with the Linodes' public IPv4 and v6 addresses? If so, would the loopback addresses not also work?

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+
+		sudo nano /etc/nginx/fastcgi_params
+
+8.  Find the line `fastcgi_param  HTTPS  $https if_not_empty;`. Below it add the following two lines, necessary for nginx to interact with the FastCGI service:
```
My default installation does not include the `fastcgi_param HTTPS $https if_not_empty;` directive in this file. Should we add it?

Fred-Zweig commented:
@alexfornuto 

> 1) Now that the minimum Linode plan has 2 gigabytes of RAM, can/should we update this variable?

Yes, I think it could be updated to 1G. So, the example memory allocation can be changed to "-s malloc,1G".

> 2) By default, the configuration file uses a lowercase m. Is there any difference as far as Varnish is concerned, or any other reason to change it here?

As the [Varnish documentation](https://www.varnish-cache.org/docs/4.0/users-guide/storage-backends.html) states, both lower case and upper case letters are equally valid when specifying the amount of memory allocated to Varnish. I chose the upper case version because it seems more clear and more frequently used.

> Are these dummy variables that should be replaced with the Linodes' public IPv4 and v6 addresses? If so, would the loopback addresses not also work?

Yes, your_server_ip should be replaced with Linode's public IPv4 address, your_server_ipv6 should be replaced with the public IPv6 address, and localhost should be left as it is because it stands for the loopback address: 127.0.0.1. In fact the word 'localhost' can be replaced with 127.0.0.1 but I think it is better to let it as localhost because Varnish will always interpret it as 127.0.0.1, or ::1 in the IPv6 sense. Being part of the "acl purger" block, the loopback address stated here as 'localhost', specifies the fact that purge requests are allowed to come from localhost.

> My default installation does not include the fastcgi_param HTTPS $https if_not_empty; directive in this file. Should we add it?

Yes, it should be added if not present. Can I add all these changes myself in a new commit now?

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+	/etc/nginx/sites-available/default
+	:   ~~~ nginx
+	
+		server {
```
Why have a server block for a 301 redirect instead of including `example-over-http.com` in the `server_name` directive of the complete server block below?

Fred-Zweig replied:
Because if we just include `example-over-http.com` in the `server_name` directive of the server block below, the website will load in any browser both as 'example-over-http.com' and as 'www.example-over-http.com', depending on what address is entered in the address bar of the browser. And we don't want that. We don't want a website to be accessible on two different addresses: one with www and the other without www. This will cause havoc in SEO and Google will penalize such a website. Any website should have only one address, be it with or without www. In my guide I described the most common scenario (and for some reasons the better scenario): that of a website that will have its address starting with www. When you want a website to have its address preceded by www, you have to expect that some users will type in a browser its address without www, so a redirect from the non-www address to the www address must be done in the server configuration. The best way to do this redirect, also the way that even the Nginx documentation recommends, is the one which I describe in my guide: two server blocks of which the first contains a 301 redirect. In fact the redirect from a www to a non-www address can be done similarly, so the users that are interested in that case can also use the details in my guide.

alexfornuto commented:
Thanks for the info @Fred-Zweig. No need for a new commit, I'm already making it part of my tech edits.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+	/etc/nginx/nginx.conf
+	:   ~~~ nginx
+	
+		# access_log /var/log/nginx/access.log;
```
Are these directives not overruled by those set in `/etc/nginx/sites-available/default`? Since we only define log paths for the https site, not the http one, aren't we leaving that site without logs? In other words, wouldn't it be safer to leave these uncommented as a catch for server blocks without log files defined?

Fred-Zweig replied:
Very interesting point. But here is the reason for my settings:
If we leave these two lines uncommented, Nginx will store in the `/var/log/nginx/access.log` file all the requests for all the https websites that we configured. It will also store all the errors for all the https websites in `/var/log/nginx/error.log`. If we have let's say three https sites configured in `/etc/nginx/sites-available/default`, Nginx will mix up all the requests to these three https websites in the access log. It will simply add the requests one below the other in the same log file: `/var/log/nginx/access.log`. It will not keep requests for different sites in different log files. We don't want this because this will make log analysis by tools such as Awstats or Webalizer impossible and even for occasional visuall checking of web traffic to each site. So, we want Nginx to write a different access log for each individual site. And the same for error logs. To do this we have to disable these general log settings and enable them in each server block for each https site individually.

For http websites things are indeed different. As I pointed out in the last paragraph of the guide, for http websites the first to receive the requests will not be Nginx but Varnish. So, there is no reason to enable Nginx access logging for http websites, either at the server block level or at the general level in `/etc/nginx/nginx.conf`, unless we want inaccurate logs. Nginx will log only the requests that will come from Varnish for those pages that were not found in cache and since most pages are found in cache, Nginx will only log a fraction of all the requests. If we want to have all the requests logged for http websites, we have to tell Varnish to log the requests for these sites. Varnish has a more complicated way of logging. It does it by starting a separate instance of a daemon called 'varnishncsa' for each individual website. The configuration of this kind of logging was too long to include in my guide. There are a few online resources that users can use to configure Varnish access logging. Since I coudn't include all this configuration, I just pointed out in the last paragraph:

> A step further can be to enable Varnish logging for the plain HTTP websites, since now Varnish is the first to receive the client requests, while nginx only receives requests for those pages that are not found in cache.

I now realize that it's good to include an error log directive in the Nginx configuration for the http site, since Varnish logs only access requests and it's recommended to also catch errors for individual http sites. So, if you have the time, please add this line before the last curly bracket in the second server block for the http website:

`error_log  /var/log/nginx/example-over-http.com.error.log notice;`

alexfornuto replied:
Thanks for the info. You'll see in my tech review (which should be done today or tomorrow) that I've moved the default log log location to match our other guides (`/var/log/www/html/example-over-http.com/log/error.log`), but I will add this line to the http server block.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -0,0 +1,683 @@
+
+23. After saving and exiting the file you have to reload the systemd process:
+
+        sudo systemctl daemon-reload
```
After running this, Varnish remains running with the previous configuration:

```
alex@afornuto-1:~$ sudo systemctl daemon-reload
alex@afornuto-1:~$ sudo systemctl status varnish
● varnish.service - Varnish HTTP accelerator
   Loaded: loaded (/lib/systemd/system/varnish.service; enabled)
   Active: active (running) since Mon 2016-11-07 16:53:46 UTC; 1h 52min ago
 Main PID: 11606 (varnishd)
   CGroup: /system.slice/varnish.service
           ├─11606 /usr/sbin/varnishd -a :6081 -T localhost:6082 -f /etc/varnish/default....
           └─11626 /usr/sbin/varnishd -a :6081 -T localhost:6082 -f /etc/varnish/default....
```

Of course, trying to restart the service at this point will fail, since nginx is already bound by default to port `80`. So I'm just wondering what's the purpose of this?

Fred-Zweig replied:
After making any changes to a .service file in `/lib/systemd/system`, the systemd process has to be reloaded with the command: `sudo systemctl daemon-reload` . If we change any .service file in `/lib/systemd/system` but don't reload the systemd process, when trying to start that service we will receive a warning like:
`Warning: Unit file of ....service changed on disk, 'systemctl daemon-reload' recommended.`
And only after reloading the systemd process the new changes that we made will take effect. So, after following the guide up to this step, after installing and configuring Varnish as the guide describes, a user must reload the systemd process with `sudo systemctl daemon-reload` in order to have Varnish start with the new configuration. But Varnish shouldn't be started yet. Please note that at this point of the guide Varnish hasn't been started. Two lines below, the guide states: `We won't start Varnish at this point because we need to configure nginx first.` So, since nginx isn't configured yet, Varnish shouldn't be started. And of course Nginx should be started either, since it hasn't been configured yet.

The output that you posted above, shows that Varnsh is already running after the `sudo systemctl daemon-reload` command, on your test server, which is not as it should be. This means you started Varnish earlier during some tests. But at this point of the guide, step 23 of the 'Install and Configure Varnish' section , after installing and configuring Varnish, as I said, Varnish is not yet started. It will be started only after configuring  and starting Nginx, in step10 of the 'Configure nginx' section.

Also, if Varnish is running properly, when checking its status with `sudo systemctl status varnish`, the CGroup lines of the output should look like:
`6189 /usr/sbin/varnishd -a :80 -T localhost:6082 -f /etc/varnish...`
Your output shows `11606 /usr/sbin/varnishd -a :6081 -T localhost:6082 -f /etc/varnish/default....` The port on which Varnsh listens, as we just set it in `/lib/systemd/system/varnish.service` should be 80 not 6081.

alexfornuto commented:
Hi @Fred-Zweig. In addition to my question on Line 394, I've staged a tech edit for you [here](https://github.com/Fred-Zweig/docs/pull/2).

It looks like I made enough formatting changes that GitHub is having a hard time showing the diff, and is instead showing me as having removed and replaced large chunks of the guide. You can try looking at the individual commits for a better view. 

Don't worry about the changes I made to our WordPress and LEMP guides, those were things I caught along the way while prepping for your guide. 

Feel free to comment on any questionable lines, just as I've done with this PR, and/or let me know when you've merged the changes.

Fred-Zweig commented:
@alexfornuto  Thank you for carefully reviewing the guide and for the important additions and edits. I merged the branch with your edits and in a second step I added a few details that I found suitable and that I will enumerate here:
- I set the title to 'How to Use Varnish with Nginx to Serve Cached Wordpress Content over SSL and HTTP on Debian 8' . I think this is a good title, since the guide equally treats the https and the http case and users must know from the title that they will find how to configure Varnish and Nginx not only for https sites but also for plain http sites.
- I decided that it's better to remove the third line from the 'Before You Begin' section, as it speaks about installing WordPress which is to be done **after** following the guide (in order to have Nginx configured first). Since the 'Before You Begin' section refers to tasks that the user must accomplish **before** following the guide, I think it's better to remove this third line which is referred to in step 12 of the 'Configure Nginx' section anyway. Nevertheless, the idea that in order to install WordPress the user has to have Nginx already running is a very good catch. I missed this order of events when first writing the guide. So, thank you for noticing this.
- I removed the `example-over-https.com` name from the two server blocks of the https site. Since the redirect from non-www to www is done in Varnish, all requests for `example-over-https.com` will be redirected to `www.example-over-https.com`. So, although it's not wrong to have both the www and non-www names in the server_name directive, I think it's better to leave just the www name, so as not to confuse the reader.
- In the 'Edit the Startup Configuration' section I added the `sudo systemctl enable varnish.service` command, because it's necessary to make Varnish start at boot up. The `START=yes` setting in `/etc/default/varnish` is not enough to make Varnish start at boot up.
- In the `Install and Configure PHP` section I added two steps in order to make sure the user has the correct settings in the `/etc/php5/fpm/pool.d/www.conf` file.
- In the 'Configure nginx' section I added a step about the user that runs Nginx, which must be `www-data`. I don't know how I forgot to mention this in the first version of the guide. Also, I mentioned an important directive that must be added to `/etc/nginx/nginx.conf`, which is ` include /etc/nginx/sites-enabled/*;` This makes Nginx read the configuration file(s) in `sites-enabled`. I also added a step for the `gzip` compression and an important step about making a symbolic link from `/etc/nginx/sites-available/default` to `/etc/nginx/sites-enabled`. I still wonder how I forgot to mention all these before.
- I also added a paragraph about a catch-all server block for the SSL protocol.

Please feel free to make any observations or changes that you find appropriate.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -310,6 +309,12 @@ For all steps in this section, replace `203.0.113.100` with your Linodes public
 
         sudo systemctl daemon-reload
 
+3.  To make Varnish start at boot up, run the following command:
```
I'm pretty sure it's enabled on install, as is the case with most services on Debian based systems.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -336,26 +341,85 @@ Before configuring nginx we have to install PHP-FPM. FPM is short for FastCGI Pr
         listen = /var/run/php5-fpm.sock
         ~~~
 
+4.  Verify that the owner of the FPM pool is 'www-data'. The `user` and `group` directives have to be set to `www-data` like this:
```
Since these settings are set like this by default, what's the purpose of these two steps?

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -336,26 +341,85 @@ Before configuring nginx we have to install PHP-FPM. FPM is short for FastCGI Pr
 ## Configure nginx
 
-1.  Open `/etc/nginx/nginx.conf` and comment out the `ssl_protocols` and `ssl_prefer_server_ciphers` directives. We'll include these SSL settings in the server block within the `/etc/nginx/sites-enabled/default` file:
+1.  Open the `/etc/nginx/nginx.conf` file. Make sure the system user that runs nginx is set to `www-data`. At the top of the file you should have the following line:
```
Same as above.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -424,15 +495,15 @@ Before configuring nginx we have to install PHP-FPM. FPM is short for FastCGI Pr
            listen  443 ssl;
            listen  [::]:443 ssl;
-           server_name  www.example-over-https.com  example-over-https.com;
+           server_name  www.example-over-https.com;
```
This worked as it was, why change it?

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -533,14 +603,20 @@ Before configuring nginx we have to install PHP-FPM. FPM is short for FastCGI Pr
     The `/var/www/html/index.html` file can contain just a simple message like "Page not found!".
 
-7.  Restart nginx, the Varnish:
+    Under the above catch-all server block you can also place a similar block for SSL traffic, including in it all the necessary SSL directives and using a self-signed SSL certificate for a dummy domain name. Otherwise, any user typing in a browser your IP address preceded by `https://`, after a browser warning, will instantly access the first HTTPS website in the list of websites configured in ` ...
```
This is unnecessary.

alexfornuto commented on docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md:
```diff
@@ -50,8 +50,6 @@ This tutorial assumes that you have SSH access to your Linode running Debian 8 (
 
 *   Follow the steps outlined in our [LEMP on Debian 8](/docs/websites/lemp/lemp-server-on-debian-8) guide. You should skip the nginx configuration section as we'll address it later on in this tutorial.
 
-*   After configuring nginx as described below, follow the steps in our [WordPress](/docs/websites/cms/how-to-install-and-configure-wordpress) guide to install and configure WordPress.
```
This is required based on my edits.

alexfornuto commented:
Hi @Fred-Zweig,

I when into the specific commits and made comments, but it essentially boils down to this: the technical changes all added unnecessary complexity to the guide, and the copy edits don't matter since another member of our docs team will conduct a full copy edit.

Additionally, changes to technical data should be complete before the tech edit is done, so that we don't have to do another complete tech test.

Please roll back these commits, and stop making changes so that we can proceed with the process.

Fred-Zweig commented:
@alexfornuto  Ok. I rolled back all the changes I made after merging your branch. I understand I was a bit overzealous with a few of my changes. I won't make changes from now on.

alexfornuto commented:
No worries! We're a little short staffed this week (it's cold season 🤒 ) but the next person available who isn't me or @brewn will grab this for copy edit.

pbzona commented:
@Fred-Zweig Thanks for your patience on this. I'll be going through and copyediting, and I'll have a PR ready within the next couple days. After that gets merged we should be in the home stretch of getting your guide published. Please feel free to reach out with any questions.

Fred-Zweig commented:
@pbzona Thank you. As it stands now, the guide has a few typos and other minor problems that I will mention below. If you have the time, when you do the final edits please also consider making the following changes:
First of all the branch has conflicts. Git signals the file  `docs/websites/lemp/lemp-server-on-debian-8.md` as a conflicting file. Probably this file has been changed before I rolled back the last 3 commits, as I was adviced.
- On line 12: The current title of the guide is: "Use Varnish with nginx to Serve Cached Wordpress Content over SSL on Debian 8". This title, that only mentions the SSL content, should be changed to reflect the real subject of the guide, namely the method to integrate Varnish with Nginx to serve cached WordPress content, for **both SSL and plain HTTP websites**. The guide equally treats the https and the http case and users must learn from the title that they will find out how to configure Varnish and Nginx not only for https sites but also for plain http sites. So, the title should be: "How to Use Varnish with Nginx to Serve Cached Wordpress Content over SSL and HTTP on Debian 8". 
- On line 378: "Begin by backup up the default server block fille:" should be changed to "Begin by making a backup of the default server block file:"
- On line 536: "Restart nginx, the Varnish:" should be changed to "Restart nginx, then start Varnish:"
- On line 298: The following sentence is obscure: "For Varnish to work we also need to edit the `varnish.service` file to use our custom configuration file, with the port number and allocated memory values updated to match the changes we made in our `/etc/default/varnish` file." To make things clear and avoid confusing the reader it should be changed to: "For Varnish to work we also need to edit the `/lib/systemd/system/varnish.service` file. We have to change the following values: the configuration file, the port number and the allocated memory. We want to make them match the values in our `/etc/default/varnish` file."
- On line 300: "Find the two lines beginning with  `ExecStart` and edit them to match the following:" should be changed to "Find the two lines beginning with `ExecStart` and make them look like this:". The word 'match' used in the first form of the sentence is misleading, especially coming after the previous sentence.
- On line 435: "server_name  www.example-over-https.com  example-over-https.com;" should be changed to "server_name  www.example-over-https.com", since this is the name of the site and the redirect made in `/etc/varnish/custom.vcl` redirects all requests for `example-over-https.com` to `www.example-over-https.com`. So there is no need for the non-www site name in this case. It isn’t a mistake but there is no need for it and I think it can confuse the reader.
- On line 468: the same as above.

pbzona commented:
@Fred-Zweig Thanks, I'll be sure to look out for these specific items and change them accordingly. I'll also see what I can do about the merge conflict, I'm sure it's just an update that we made after your initial commit. I should have a PR ready today or tomorrow morning at the latest.

pbzona commented:
@Fred-Zweig I've made my PR here: https://github.com/Fred-Zweig/docs/pull/3

Please review it at your convenience. I *think* this should resolve the merge conflict, but if it doesn't I'll take another look. Very nice job with this guide, can't wait to try it out on some of my own servers.

Not to veer too far off-topic, but have you tried this with newer versions of nginx? I don't intend to change the installation method in your guide, but I normally install nginx from backports when I run it on Debian (for better HTTP2 compatibility mostly), so I'm curious if there would be major differences.

Fred-Zweig commented:
@pbzona Thank you for the edits. I merged the PR. It seems the merge conflict is still present.

Indeed, the settings I describe in my guide work with all the newer versions of Nginx since these settings use the core functionality of Nginx which hasn't been changed. I also wanted to enable HTTP2 and I installed the newest version of Nginx from their official website, Nginx 1.11.1 (a few months ago). I also tried a previous version, if I remember correctly 1.10.2, and in all cases it worked perfectly. I even compiled Nginx 1.11.1 from source to add the ngx_pagespeed module and it worked perfectly, with and without the module enabled. So, I'm sure the current backports version of Nginx, which is 1.9.10, being older than the versions I tested, has no reason not to work with the exact settings in my guide.

pbzona commented:
@Fred-Zweig Very cool, I'll definitely be using this for some personal projects in that case.

I've made you an additional PR, sorry for the double. I replaced your diagram with one using Linode-styled graphics, but forgot to remove the originals. Please review and merge when you can: https://github.com/Fred-Zweig/docs/pull/4

As for the merge conflict, it seems the easiest way will be to resolve it when we're ready to merge, so don't worry about it for now. We'll handle that once it's ready to publish. Thanks!

alexfornuto commented:
@Fred-Zweig @pbzona Sorry about the merge conflict, that was me. I made a personal note of it, but forgot to mention it for others working on this guide. It's a simple fix I made on the LEMP guide, that conflicts with a formatting adjustment upstream. We'll take care of it during merge, easy-peasy.

Fred-Zweig commented:
@alexfornuto Ok. Thank you.
@pbzona The new scheme picture looks great. Yet I noticed the following line in it: 'Varnish replies **to to** Nginx data retrieved from cache' . Also, I think the two arrows coming from the clients can be double sense arrows to indicate that data travels not only from the clients to Nginx/Varnish but also backwards, from Nginx/Varnish to the clients. I can edit the picture myself and add it to a new commit if needed.

alexfornuto commented:
Hey @Fred-Zweig, I was working on a P.S. when I saw your post. So let me make a few points here.

1) Thanks for pointing out the issues with the graphic. We'll have our graphics artists make the corrections. Please go ahead and merge Phil's latest PR, and we can take it from there.

2) We need your bounty payment info. You can receive payment through PayPal or Linode credit. In either case we need the email address of the account. If you don't want to post it here, you can email contribute@linode.com. If you do, make sure to specify which guide this is for. 

3) Just a heads up, while we make the final changes and touch-ups on this guide, the title may change. It's a mouthful, and while a WordPress site is used as an example (and includes info about WP plugins), the guide is really CMS independent and could be followed with out without WordPress.

Fred-Zweig commented:
@alexfornuto Thank you. I merged the last PR and I sent my payment info to contribute@linode.com .

Your point about the fact that the solution proposed by the guide is CMS independent is correct, but here is the explanation for my approach:

I read an enormous amount of tutorials that give just very general settings that can't be actually used in a real-life situation without adding tons of other settings that can be figured out only after reading tons of additional documentation. So, what's the use of such guides. If a guide pretends to give step by step instructions and promises that at the end you'll have a really working setup, then why after following the guide one has to read other tons of guides or official documentation to get things working ? So, I decided that any guide I will write will give users really complete instructions so as at the end they will have a really working and thus a really useful setup. We all know that all admins and developers use some kind of CMS today. So, a guide about Nginx and Varnish settings that doesn't address the specific requirements of a CMS, which is always used in real-life situations, will be an incomplete guide. It will leave readers with a big question: 'What other settings do I have to make for my specific CMS ?'. So, as I wasn't able to write specific settings for all CMS's I chose the most popular and the best of all: WordPress; and I wrote the specific WordPress settings for Varnish and about the plugin needed for purges. So, although in essence my guide is CMS independent, in order to be considered a complete guide, it has to address the specific settings needed by a particular CMS. Of course, people that use other CMS, such as Joomla or Drupal can adapt the settings to their situation, by leaving aside those WordPress related Varnish settings or the WordPress plugin, and by adding other specific settings that they will have to figure out by themselves.

So, I think that as it is now, the title is appropriate as it is complete and it reflects the real subject of the guide. It is indeed too long, and I would gladly propose to change it to a shorter one, but can somebody devise a shorter title that would convey the same message ?

And by the way, the word 'Nginx' in the title has been changed to 'nginx'. I know that the Linode standards in writing Nginx impose that 'nginx' will be the name of the program and 'Nginx' the name of the company, but in this case, in which the word is part of the title, it can be written as 'Nginx'. We write with capital letters words like 'Serve' and 'Content' and we can't write 'Nginx' ?

alexfornuto commented:
The open-source nginx project is always lowercase, even when in a title or the start of a sentence. For reference: https://nginx.org/en/

Fred-Zweig commented:
@alexfornuto You are right but the nginx developers don't seem to be always consistent with the way they write nginx. The open-source server nginx is often written as 'NGINX', as it can be seen [here](https://www.nginx.com/resources/wiki/start/topics/tutorials/install/) or [here](https://www.nginx.com/resources/wiki/community/faq/#how-do-you-pronounce-nginx). This is the open-source server, not the paid version written as 'NGINX Plus'. I'm aware that the two links I just mentioned are from `nginx.com` and not `nginx.org`, but the articles are written by the nginx developers and they do refer to the open-source server that can be downloaded as a Linux package, not to the paid version 'NGINX Plus'.  Although I wouldn't consider Wikipedia an authoritative source regarding the way proper names are written, it also shows the ['Nginx' writing](https://en.wikipedia.org/wiki/Nginx). The majority of the Linode guides themselves, use the 'Nginx' writing, both in the title and inside the guide, as can be seen [here](https://www.linode.com/docs/websites/nginx/how-to-configure-nginx) or [here](https://www.linode.com/docs/websites/nginx/installing-nginx-on-ubuntu-12-04-lts-precise-pangolin). There are indeed Linode guides that use the 'nginx' writing, like [this one](https://www.linode.com/docs/websites/nginx/install-nginx-and-a-startssl-certificate-on-debian-8-jessie). In total, there seem to be more guides with the 'Nginx' writing than with the 'nginx' writing, as can be seen [here](https://www.linode.com/docs/websites/nginx/).

Anyway, I don't consider the way nginx is written as a real problem. Since my guide has been edited to have only 'nginx' inside the guide, I can agree to also have it written as 'nginx' in the title.

But for the title of the guide to be accurate, it has to contain 6 references: Varnish, Nginx, WordPress, HTTP, SSL and Debian 8. So, I think that the current title which mentions all the 6 elements, although a bit long, is a good title.

alexfornuto commented:
Congrats @Fred-Zweig, your guide has been [published](https://linode.com/docs/websites/varnish/use-varnish-and-nginx-to-serve-wordpress-over-ssl-and-http-on-debian-8)! You should receive your bounty promptly. Thanks for contributing!

Fred-Zweig commented:
@alexfornuto , @pbzona , @brewn  Thank you for the time and effort you put in publishing this guide.
I apologize if I was at times difficult to work with. I insisted so much on some details just because I wanted the guide to be published in the best form so as to benefit the readers. And in this spirit of attention to detail please let me make a few final observations:

- In the 'Before You Begin' section, in the third entry, there is the following typo: 'We’ll include a step in the instructions **to let you know you** when it’s time to do this.'
- In the 'More Information' section, the 'NGINX Documentation' entry  has the Varnish documentation URL linked to it, instead of the Nginx documentation URL which is: `http://nginx.org/en/docs/`
- In the 'Install and Configure Varnish' section, in the third entry, in the second paragraph, the `-s malloc,1g` setting is mentioned. This setting appears in two other places as `-s malloc,1G`, with capital G. Although `G` and `g` are equally valid for RAM memory specification in Varnish, for consistency reasons I think it's better to write `-s malloc,1G` in this instance also.
- In the section 'How Varnish and nginx Work Together', at the end of the second paragraph there is the following sentence: 'then Varnish will store the fetched content in cache and deliver it to the client on port 80.' This wording seems correct but I noticed that in all the paragraphs where similar reference to cache is made, the word 'cache' has been replaced with 'the cache'. So, for consistency reasons I think in this instance also 'cache' should be replaced with 'the cache'.
- I attached the picture with the Varnish scheme edited, without the typo in 'Varnish replies **to to** Nginx data retrieved from cache' and with double sense arrows coming from the clients. The picture has the same resolution (649 x 626), name (varnish-cache) and type (.png) as the one currently included in the guide. So, you can just replace the 'imperfect' picture with this picture:

![varnish-cache](https://cloud.githubusercontent.com/assets/17623443/20581125/79a585e6-b1df-11e6-821f-187e7b4d804d.png)

P.S. I can confirm that I received the payment through PayPal, so thanks again @alexfornuto and Linode.

P.P.S. The setup described in this guide is even 'scalable' in the sense that it's very efficient when used in common situations where admins want to serve cached WordPress content over HTTP and SSL for a small number of sites hosted on the same server. This efficiency is due to the power and versatility of Nginx, the same instance of Nginx being capable to decrypt SSL traffic and serve HTTP content simultaneously. But for a big number of sites hosted on the same server, the setup can be 'scaled up' by using a separate instance of Nginx installed on a separate physical server or VPS for SSL decryption. So, there will be one physical server or VPS with frontend Nginx and a second physical server or VPS with Varnish and backend Nginx. This way the two instances of Nginx will have lots of physical resources to easily handle any traffic for any big number of sites. Of course in this case the `proxy_pass` directive in the server block of frontend Nginx will be changed from `proxy_pass http://127.0.0.1:80;` to `proxy_pass http://ip-of-backend-nginx:80;`
